src/java.base/share/classes/java/io/SerializationMisdeclarationChecker.java

```diff
@@ -0,0 +1,300 @@
+    private static final String WRITE_REPLACE_NAME = "writeReplace";
+    private static final String READ_RESOLVE_NAME = "readResolve";
+
+    private static final Class<?>[] WRITE_OBJECT_PARAM_TYPES = {ObjectOutputStream.class};
+    private static final Class<?>[] READ_OBJECT_PARAM_TYPES = {ObjectInputStream.class};
+    private static final Class<?>[] READ_OBJECT_NO_DATA_PARAM_TYPES = {};
+    private static final Class<?>[] WRITE_REPLACE_PARAM_TYPES = {};
+    private static final Class<?>[] READ_RESOLVE_PARAM_TYPES = {};
+
+    private final Class<?> cl;
+
+    SerializationMisdeclarationChecker(Class<?> cl) {
+        this.cl = cl;
+    }
+
+    @SuppressWarnings("removal")
+    void checkMisdeclarations() {
+        AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+            privilegedCheckSerialVersionUID();
+            privilegedCheckSerialPersistentFields();
+
+            privilegedCheckPrivateMethod(WRITE_OBJECT_NAME, WRITE_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(READ_OBJECT_NAME, READ_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(READ_OBJECT_NO_DATA_NAME, READ_OBJECT_NO_DATA_PARAM_TYPES, Void.TYPE);
+
+            privilegedCheckAccessibleMethod(WRITE_REPLACE_NAME, WRITE_REPLACE_PARAM_TYPES, Object.class);
+            privilegedCheckAccessibleMethod(READ_RESOLVE_NAME, READ_RESOLVE_PARAM_TYPES, Object.class);
+
+            return null;
+        });
+    }
+
+    private void privilegedCheckSerialVersionUID() {
+        Field f = declaredField(cl, SUID_NAME);
+        if (f == null) {
+            if (isOrdinaryClass()) {
+                commitEvent(SUID_EXPLICIT,
+                        "a private static final long " + SUID_NAME +
+                                " field should be declared");
+            }
+            return;
+        }
+        if (cl.isEnum()) {
+            commitEvent(SUID_INEFFECTIVE_ENUM,
+                    SUID_NAME + " is not effective in an enum class");
+        }
+        if (!isPrivate(f)) {
+            commitEvent(SUID_PRIVATE,
+                    SUID_NAME + " should be declared private");
+        }
+        if (!isStatic(f)) {
+            commitEvent(SUID_STATIC,
+                    SUID_NAME + " must be declared static to be effective");
+        }
+        if (!isFinal(f)) {
+            commitEvent(SUID_FINAL,
+                    SUID_NAME + " must be declared final to be effective");
+        }
+        if (f.getType() != Long.TYPE) {
+            commitEvent(SUID_LONG,
+                    SUID_NAME + " should be declared of type long");
+            if (!isStatic(f)) {
+                return;
+            }
+            f.setAccessible(true);
+            if (longFromStatic(f) == null) {
+                commitEvent(SUID_CONVERTIBLE_TO_LONG,
+                        SUID_NAME + " must be convertible to long via widening to be effective");
```

Comment:
The serialization spec only shows using long.  If any recommendation is made it should be to declare the field as a `long`

Reply:
There's a check on the type at L.104 which is about the "should" recommendation, since serialization does not care about the type of the field being `long`.

This check is about the value at runtime, which is a "must" because serialization expects it to be convertible to long by widening.

Reply:
The implementation should not show through. I don't remember if the particular implementation was intentional or just a shortcut to get the value. But any suggestion that implies a fix should be to recommend the best practice.